File: merge-intervals/interval-list-intersections.js
import { Interval } from "./interval.js";

// Function to find the intersecting points between two intervals
export function intervalsIntersection(intervalListA, intervalListB) {
	let result = [];
	let a = 0;
	let b = 0;

	while (a < intervalListA.length && b < intervalListB.length) {
		const intA = intervalListA[a];
		const intB = intervalListB[b];

		const start = Math.max(intA.start, intB.start);
		const end = Math.min(intA.end, intB.end);

		if (start <= end) {
			result.push(new Interval(start, end));
		}

		if (intA.end < intB.end) {
			a++;
		} else {
			b++;
		}
	}
	return result;
}

File: merge-intervals/interval-list-intersections.py
from interval import Interval


# Function to find the intersecting points between two intervals
def intervals_intersection(interval_list_a, interval_list_b):
    result = []
    a = b = 0

    # end once we've reached the end of either list
    while a < len(interval_list_a) and b < len(interval_list_b):
        i_a, i_b = interval_list_a[a], interval_list_b[b]

        start = max(i_a.start, i_b.start)  # get the later start time
        end = min(i_a.end, i_b.end)  # get the earlier end time

        if start <= end:  # if the times overlap, the intervals overlap
            result.append(Interval(start, end))  # add the intersection

        # we can now move on from the earlier ending interval
        if i_a.end < i_b.end:
            a += 1
        else:
            b += 1

    return result
